packages/react/src/utils/mergeThemes.ts

```diff
@@ -36,12 +36,38 @@ export const emptyTheme: ThemePrepared = {
   staticStyles: [],
   icons: {},
   animations: {},
+  hash: '',
 }
 
 // ----------------------------------------
 // Component level merge functions
 // ----------------------------------------
 
+const mergeComponentStylesCache = {}
+const mergeComponentStylesThemeCache = {}
+
+export const mergeComponentStylesWithCache = (
+  theme,
+  hashObj,
+  sources: (ComponentSlotStylesInput | null | undefined)[],
+) => {
+  try {
+    const themeHash = theme.hash || JSON.stringify(theme)
+    if (!mergeComponentStylesThemeCache[themeHash]) {
+      mergeComponentStylesThemeCache[themeHash] = Object.keys(mergeComponentStylesThemeCache).length
+    }
+
+    const hashVal = JSON.stringify({ ...hashObj, theme: mergeComponentStylesThemeCache[themeHash] })
```

Comment:
I think we should try a WeakMap here instead.  You can use a complex object for the key in that case and avoid the hashing and JSON.stringify.

We want to cache on the theme object and a filtered set of component props, the props that are used in the styles.  If the theme is the same, and the props used in the styles are the same, then we know we can return the cached result.

This way, we only have to merge, resolve, and render styles for new themes and prop pairs that we haven't seen before.

Reply:
I don't see how we can use WeakMap for the caching, as it is comparing based on the reference:
```
// theme1 and theme2 are basically the same object, but different references

const theme1 = {
  backgroundColor: 'red',
  color: 'white',
}

const theme2 = {
  backgroundColor: 'red',
  color: 'white',
}

const map = new WeakMap();
map.set(theme1, "Theme1");

if(map.has(theme2)) { // we would want this to return true, but it returns false
  console.log("equal") 
}

//adding new prop
theme1.fontSize = "14px"

if(map.has(theme1)) { // more over we definitelly don't want this to happen, but it does as it is the same reference
  console.log("Still has theme1")
}
```

I understand that JSON.stringify is not that cheap and I tried the `object-hash` package, but it is even slower... When comparing we have to be sure that the structure of the object are the same, not their reference. If I am missing something, or you have other suggestion let me know.

Reply:
Updating from offline chat:

The difference here is that `theme1 === theme2` is lightning fast and simple to implement.  The cost of deep checking if one object equals another or maintaining a hash/id algorithm will be much less performant.

So, let's shift the complexity of this problem from hashing/ids to ensuring immutable themes.